### PR TITLE
feat: [IDP-1137]: Helm chart support for idp-service

### PR DESCRIPTION
**Pull request by devesh-harness**
## Describe your changes
-> Helm chart support for idp-service for on boarding idp-service to devsapces.
-> change in build_jar.sh to support building of idp-service from common build pipeline for devspaces.
-> TEST - with these config changes we are able to successfully build and deploy the idp-service in devspaces.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/53079)

### CODE REVIEW

**Review by vikyathharekal — DISMISSED** [review]

**Comment by arya-harness** [comment]
it is missing virtualservice.yaml

**Review by harness-jc — APPROVED** [review]

**Review by harness-jc — CHANGES_REQUESTED** [review]
please create virtualservice.yaml to support istio installations.

File: idp-service/build/container/Dockerfile-idp-service-multistage [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  /app/dist/idp-service/profile.jfc \
+  /app/dockerization/base-images/apm/*.sh \
+  /opt/harness/
+RUN chmod +x /opt/harness/*.sh
+CMD [ "/opt/harness/run.sh" ]
+
+############################ON PREM#########################
+FROM base as onprem
+
+RUN /opt/harness/inject-onprem-apm-bins-into-dockerimage.sh && rm /opt/harness/inject-onprem-apm-bins-into-dockerimage.sh
```
[the remark this answers is not in the document]
Comment (devesh-harness) [reply idp-service/build/container/Dockerfile-idp-service-multistage:45]: as discussed we will make these changes when we will do for all services.

File: idp-service/chart/Chart.yaml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+# to be deployed.
+#
+# Library charts provide useful utilities or functions for the chart developer. They're included as
+# a dependency of application charts to inject those utilities and functions into the rendering
+# pipeline. Library charts do not define any templates and therefore cannot be deployed.
+type: application
+# This is the chart version. This version number should be incremented each time you make changes
+# to the chart and its templates, including the app version.
+# Versions are expected to follow Semantic Versioning (https://semver.org/)
+version: 0.13.0
```
[the remark this answers is not in the document]
Comment (devesh-harness) [reply idp-service/chart/Chart.yaml]: as discussed changed it.

**Comment by devesh-harness** [comment]
> please create virtualservice.yaml to support istio installations.
done please check now.

**Review by harness-jc — APPROVED** [review]

**Comment by devesh-harness** [comment]
trigger uts

**Comment by devesh-harness** [comment]
trigger ut8

**Review by vikyathharekal — APPROVED** [review]

File: idp-service/chart/templates/_helpers.tpl [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+{{/*
+Create the name of the security image to use
+*/}}
+{{- define "idp-service.securityImage" -}}
+{{ include "common.images.image" (dict "imageRoot" .Values.securityImage.image "global" .Values.global) }}
+{{- end }}
+
+{{/*
+Create the name of the s3 image to use
```
Comment (arya-harness) [inline idp-service/chart/templates/_helpers.tpl:73]: This is not required

File: idp-service/chart/templates/_helpers.tpl [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+{{- end }}
+{{- define "idp-service.addOnImage" -}}
+{{ include "common.images.image" (dict "imageRoot" .Values.addOnImage.image "global" .Values.global) }}
+{{- end }}
+{{- define "idp-service.leImage" -}}
+{{ include "common.images.image" (dict "imageRoot" .Values.leImage.image "global" .Values.global) }}
+{{- end }}
+
+
+{{- define "idp-service.pullSecrets" -}}
```
Comment (arya-harness) [inline idp-service/chart/templates/_helpers.tpl:86]: this function is not required

File: idp-service/chart/templates/deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+          imagePullPolicy: {{ .Values.image.pullPolicy }}
+          envFrom:
+            - configMapRef:
+                name: idp-service
+          ports:
+            - name: http
+              containerPort: {{.Values.service.port}}
+              protocol: TCP
+          env:
+            {{- include "harnesscommon.dbconnection.redisEnv" (dict "context" .Values.global.database.redis "userVariableName" "EVENTS_FRAMEWORK_REDIS_USERNAME" "passwordVariableName" "EVENTS_FRAMEWORK_REDIS_PASSWORD") | indent 12 }}
```
Comment (arya-harness) [inline idp-service/chart/templates/deployment.yaml:82]: does idp support redis sentinels?

File: idp-service/chart/templates/ingress.yaml [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+    nginx.ingress.kubernetes.io/rewrite-target: /$1
+spec:
+  {{- if and .Values.global.ingress.className (semverCompare ">=1.18-0" .Capabilities.KubeVersion.GitVersion) }}
+  ingressClassName: {{ .Values.global.ingress.className }}
+  {{- end }}
+  tls:
+  rules:
+    - http:
+        paths:
+          - path: {{.Values.pathPrefix}}/(v1/backstage-env-variables.*)
```
Comment (arya-harness) [inline idp-service/chart/templates/ingress.yaml:43]: remove {{.Values.pathPrefix}}
/(v1/backstage-env-variables.*) is enough

Comment (arya-harness) [reply idp-service/chart/templates/ingress.yaml:43]: lets do it for all openapi specs

File: idp-service/chart/templates/virtualservice.yaml [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  name: {{ include "idp-service.name" . }}
+  namespace: {{ .Release.Namespace }}
+  labels:
+    app.kubernetes.io/instance: {{ .Release.Name }}
+    {{- if .Values.global.commonLabels }}
+    {{- include "harnesscommon.tplvalues.render" ( dict "value" .Values.global.commonLabels "context" $ ) | nindent 4 }}
+    {{- end }}
+  annotations:
+    {{- if .Values.virtualService.annotations }}
+      {{- include "harnesscommon.tplvalues.render" (dict "value" .Values.ingress.annotations "context" $) | nindent 4 }}
```
Comment (arya-harness) [inline idp-service/chart/templates/virtualservice.yaml:14]: {- if .Values.virtualService.annotations }}
      {{- include "harnesscommon.tplvalues.render" (dict "value" .Values.ingress.annotations "context" $) | nindent 4 }}
      
      needs to be changed to
      {- if .Values.virtualService.annotations }}
      {{- include "harnesscommon.tplvalues.render" (dict "value" .Values.virtualService.annotations "context" $) | nindent 4 }}

File: idp-service/chart/values.yaml [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+  imagePullSecrets: []
+waitForInitContainer:
+  image:
+    registry: docker.io
+    repository: harness/helm-init-container
+    pullPolicy: IfNotPresent
+    tag: "latest"
+    digest: ""
+    imagePullSecrets: []
+securityImage:
```
Comment (arya-harness) [inline idp-service/chart/values.yaml:110]: we can remove unused values.

**Review by jyotiraval — APPROVED** [review]
Placeholder seceret

File: idp-service/chart/templates/_helpers.tpl (hunk 3 above)
Comment (devesh-harness) [reply idp-service/chart/templates/_helpers.tpl]: done

File: idp-service/chart/templates/_helpers.tpl (hunk 4 above)
Comment (devesh-harness) [reply idp-service/chart/templates/_helpers.tpl]: done

File: idp-service/chart/templates/deployment.yaml [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+      initContainers:
+      {{- if .Values.global.database.mongo.installed }}
+      - name: wait-for-mongo
+        image: {{ include "common.images.image" (dict "imageRoot" .Values.waitForInitContainer.image "global" .Values.global) }}
+        imagePullPolicy: IfNotPresent
+        args:
+          - "pod"
+          - "-lapp=mongodb-replicaset"
+      {{- end }}
+      {{- if .Values.global.database.timescaledb.installed }}
```
[the remark this answers is not in the document]
Comment (devesh-harness) [reply idp-service/chart/templates/deployment.yaml]: done

File: idp-service/chart/templates/ingress.yaml (hunk 6 above)
Comment (devesh-harness) [reply idp-service/chart/templates/ingress.yaml]: done

File: idp-service/chart/templates/virtualservice.yaml (hunk 7 above)
Comment (devesh-harness) [reply idp-service/chart/templates/virtualservice.yaml]: done

File: idp-service/chart/templates/ingress.yaml [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+                  number: 12003
+          - path: {{.Values.pathPrefix}}/(v1/entity-facets.*)
+            pathType: Prefix
+            backend:
+              service:
+                name: {{.Values.name}}
+                port:
+                  number: 12003
+---
+apiVersion: networking.k8s.io/v1
```
[the remark this answers is not in the document]
Comment (devesh-harness) [reply idp-service/chart/templates/ingress.yaml]: removed it for now.

**Comment by devesh-harness** [comment]
trigger uts

File: idp-service/chart/templates/_helpers.tpl [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+{{- define "idp-service.serviceAccountName" -}}
+{{- if .Values.serviceAccount.create }}
+{{- default (include "idp-service.fullname" .) .Values.serviceAccount.name }}
+{{- else }}
+{{- default "default" .Values.serviceAccount.name }}
+{{- end }}
+{{- end }}
+
+
+{{/*
```
Comment (arya-harness) [inline idp-service/chart/templates/_helpers.tpl:65]: We can remove securityImage too, IDP does not require this I believe.

**Review by arya-harness — DISMISSED** [review]

**Comment by devesh-harness** [comment]
trigger uts

**Comment by devesh-harness** [comment]
trigger uts

**Comment by devesh-harness** [comment]
trigger ut7

**Comment by devesh-harness** [comment]
trigger ut7

**Review by jyotiraval — APPROVED** [review]

**Review by arya-harness — APPROVED** [review]

**Review by harness-jc — APPROVED** [review]

**Comment by devesh-harness** [comment]
trigger review

**Review by sathish-soundarapandian — APPROVED** [review]

**Comment by devesh-harness** [comment]
trigger review

**Comment by devesh-harness** [comment]
trigger review

**Pull request merged**